temba/ussd/models.py

```diff
@@ -117,8 +117,19 @@ def handle_incoming(cls, channel, urn, date, external_id, contact=None, message_
         # check if there's an initiated PUSH session
         session = cls.objects.get_initiated_push_session(contact)
 
+        created = False
         if not session:
-            session, created = cls.objects.update_or_create(external_id=external_id, defaults=defaults)
+            try:
+                session = cls.objects.select_for_update().exclude(status__in=ChannelSession.DONE)\
```

Comment:
@nicpottier I see the `update_or_create` function we replaced also does a `select_for_update` so it's not adding a new lock: https://github.com/django/django/blob/master/django/db/models/query.py#L469